File: src/go/4_BehavioralPatterns/6_Memento/memento.go
package memento

import (
  "fmt"
)

type Memento struct {
  Balance int
}

type BankAccount struct {
  balance int
}

func (b *BankAccount) Deposit(amount int) *Memento {
  b.balance += amount
  return &Memento{b.balance}
}

func (b *BankAccount) Restore(m *Memento) {
  b.balance = m.Balance
}

func main() {
  ba := BankAccount{100}
  m1 := ba.Deposit(50)
  m2 := ba.Deposit(25)
  fmt.Println(ba)

  ba.Restore(m1)
  fmt.Println(ba) // 150

  ba.Restore(m2)
  fmt.Println(ba)
}

File: src/java/4_BehavioralPatterns/6_Memento/Memento.java
package com.activemesa.behavioral.memento;

class Memento
{
  public int balance;

  public Memento(int balance)
  {
    this.balance = balance;
  }
}

class BankAccount
{
  private int balance;

  public BankAccount(int balance)
  {
    this.balance = balance;
  }

  public Memento deposit(int amount)
  {
    balance += amount;
    return new Memento(balance);
  }

  public void restore(Memento m)
  {
    balance = m.balance;
  }

  @Override
  public String toString()
  {
    return "BankAccount{" +
      "balance=" + balance +
      '}';
  }
}

class MementoDemo
{
  public static void main(String[] args)
  {
    BankAccount ba = new BankAccount(100);
    Memento m1 = ba.deposit(50); // 150
    Memento m2 = ba.deposit(25); // 175
    System.out.println(ba);

    // restore to m1
    ba.restore(m1);
    System.out.println(ba);

    // restore to m2
    ba.restore(m2);
    System.out.println(ba);
  }
}